hera_pspec/grouping.py

```diff
@@ -288,74 +308,88 @@ def average_spectra(uvp_in, blpair_groups=None, time_avg=False,
                     if store_cov:
                         cov = uvp.get_cov((spw, blp, p))
-                    # Get error bar weights and set invalid ones to zero.
+                        # shape of cov: (Ntimes, Ndlys, Ndlys)
+                    # Get error bar
                     errws = {}
                     for stat in stat_l:
-                        errs = uvp.get_stats(stat, (spw, blp, p)).copy()
-                        no_data = np.isnan(errs)
-                        errs[~no_data] = 1. / errs[~no_data] ** 2.
-                        errs[no_data] = 0.
-                        errws[stat] = errs
-
+                        errws[stat] = (uvp.get_stats(stat, (spw, blp, p)))**2
+                        # shape of errs: (Ntimes, Ndlys)
+
+                    if use_error_weights:
+                    # If use_error_weights==True, all arrays are weighted by a specified kind of errors,
+                    # including the error_filed in stats_array and cov_array. 
+                    # For each power spectrum P_i with error sigma_i, 
+                    # P_avg = \sum{ P_i / (sigma_i)^2 } / \sum{ 1 / (sigma_i)^2 } 
+                    # while for terms like variance or covariance,
+                    # epsilon_avg = \sum{ (epsilon_i / (sigma_i)^4 } / ( \sum{ 1 / (sigma_i)^2 } )^2
+                    # For reference: M. Tegmark 1997, The Astrophysical Journal Letters, 480, L87, Table 1, #3
+                    # or J. Dillon 2014, Physical Review D, 89, 023002 , Equation 34. 
+                        w = 1./(uvp.get_stats(error_weights, (spw, blp, p)))**2
+                        # shape of w: (Ntimes, Ndlys)
+                    else:
+                    # Otherwise all arrays are averaged in a way weighted by the integration time,
+                    # including the error_filed in stats_array and cov_array. 
+                        w = (ints * np.sqrt(nsmp))
+                        # shape of w: (Ntimes, 1)
+                    
                     # Take time average if desired
                     if time_avg:
-                      data = (np.sum(data * w, axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
-                      wgts = (np.sum(wgts * w[:, None], axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf)[:, None])[None]
-                      ints = (np.sum(ints * w, axis=0) \
-                           / np.sum(w, axis=0).clip(1e-10, np.inf))[None]
-                      nsmp = np.sum(nsmp, axis=0)[None]
-                      w = np.sum(w, axis=0)[None]
-                      if store_cov:
-                          cov = (np.sum(cov * w[:, None], axis=0) \
-                          / np.sum(w,axis=0).clip(1e-10, np.inf))[None]
-
-                      for stat in stat_l:
-                        errws[stat] = np.sum(errws[stat], axis=0)[None]
+                        data = (np.sum(data * w, axis=0) \
```

Comment:
why are the `.clip` methods being removed? they are important for weights that are all zeros (which can happen)

Reply:
I remove the line since I found a cut-off at 1e-10 is not appropriate when we use inverse variance as weights, for they could be as small as 1e-20.

Reply:
Okay, then let's keep the `clip` methods and just change the tolerance to `1e-20`.

Reply:
To be more cautious, I set the tolerance as 1e-40.